### DISCUSSION TITLE
[Feature]: Vue Support

### DISCUSSION THREAD

**Discussion by atilkan** [remix-run/remix#2899] [Proposals]
### What is the new or updated feature that you are suggesting?

Support Vue.js

### Why should this feature be included?

The same reason for React, Preact.

**Comment by egoist** [comment]
This is definitely out of the scope of this project 🥶

**Comment by sergiodxa** [comment]
> This is definitely out of the scope of this project 🥶

I don’t think so, Remix uses React but it’s not really tied to it, if you see there’s a package with the related React code, so there could be another with Vue, and the same way Remix let you choose between Node or CF Workers, you could choose between React or Vue

**Comment by patrickcurl** [comment]
I agree with what @sergiodxa said... if you think about it...a framework that's a bit like next or remix that is frontend agnostic with simple hooks to basically interop between react, vue, svelte, alpine, etc would be a real game changer...

it'd be like if next and nuxt were one and the same with the same API just with options on vue SFCs or react components... instead of learning something new, there'd be a unified standard....

in a way I think inertiajs sort of does this at least for laravel and vue apps.. but I like the app structure with remix and it feels snappier than laravel+inertiajs+vue...

**Comment by Flexicon** [comment]
I'll chime in here as well with the fact that Remix could have the potential of being something like [Astro](https://astro.build/) but for actual apps, since Astro pretty much aims at building SSG websites.

**Comment by MichaelDeBoey** [comment]
I kind of agree with @sergiodxa on here, but I also see problems arising here:
Remix is built upon `react-router`, which is (obviously) a React library.

Supporting Vue would mean that we would need an API compatible `vue-router` kind of thing, which is not something the Remix team wants to maintain I think.
A third party could perfectly create a `remix-vue` though and the Remix team can point to that as sort of an endorsement, but I think that if we're being realistic that official (real) support will not happen as long as it's not officially maintained by the Remix team.

Nothing forbids the team to hire a Vue expert though 🤷‍♂️

**Comment by jeff-cc** [comment]
I just started [remix-vue](https://github.com/remix-vue/remix-vue)

**Comment by patrickcurl** [comment]
> I kind of agree with @sergiodxa on here, but I also see problems arising here:
> Remix is built upon react-router, which is (obviously) a React library.
It'd be probably a diff library altogether, but maybe swapping react-router for inertiajs which supports react/vue but maybe lacks things needed for remix to work, IDK.... 

Honestly, a framework agnostic router would be tight and might be a good project unto itself, for anyone w/ the time/ability to focus on that...

**Comment by Brendan-csel** [comment]
I'm just going to tag [SolidJs](https://www.solidjs.com/) onto this issue.  It already has a [router](https://github.com/solidjs/solid-app-router) with API very similar to react-router 6.

**Comment by jacob-ebey** [comment]
> I kind of agree with @sergiodxa on here, but I also see problems arising here: Remix is built upon `react-router`, which is (obviously) a React library.
> 
> Supporting Vue would mean that we would need an API compatible `vue-router` kind of thing, which is not something the Remix team wants to maintain I think. A third party could perfectly create a `remix-vue` though and the Remix team can point to that as sort of an endorsement, but I think that if we're being realistic that official (real) support will not happen as long as it's not officially maintained by the Remix team.
> 
> Nothing forbids the team to hire a Vue expert though 🤷‍♂️

`react-router` actually has no dependency on react and is just a matching algorithm when it comes down to it and is something that 100% could be integrated into a vue, svelte, solid, etc, router. Remix itself is framework agnostic as well consisting of a server runtime based on fetch, and a client transition manager that is based on state updates that can be wired into any frontend framework. I'm excited myself to explore frameworks other than react and believe we are uniquely positioned to achieve this with relative ease compared to others.

**Comment by MichaelDeBoey** [comment]
> `react-router` actually has no dependency on react and is just a matching algorithm when it comes down to it and is something that 100% could be integrated into a vue, svelte, solid, etc, router.

`react-router` does have a peerDependency on `react` though
https://github.com/remix-run/react-router/blob/48b7c60a2133a617d64533b54eb299dbd2d1be0e/packages/react-router/package.json#L16-L18

**Comment by HerrBertling** [comment]
I guess for documentation purposes, this should be linked here: https://www.npmjs.com/package/remix-router-vue 😄

**Reply by Xiphe** [reply]
Hallo Markus 👋❤️

**Comment by binajmen** [comment]
Sooo. I might have a mission where Vue is used, and would love to bring Remix in the equation. How far are we in the rule-'em-all process? 😄

**Comment by brophdawg11** [comment]
<img width="219" alt="Screenshot 2022-12-19 at 12 46 43 PM" src="https://user-images.githubusercontent.com/1609022/208487936-f65d86e6-989a-4042-a493-85cc822b0294.png">

😂 23 emoji reactions yet only 5 upvotes.  Just a heads up to all the folks that reacted with 👍  or 👀  on the proposal here - the upvote (↑) is all that counts towards raising visibility so don't forget to upvote it if this is something you'd like to see worked on!

**Reply by ThaDaVos** [reply]
Is the upvote arrow new? Wasn't there when I voted on it with 👍

**Reply by brophdawg11** [reply]
Yeah it's possible it was just an issue back in Nov 2021 and since been converted to a Discussion

**Comment by expressodaily** [comment]
An important piece of information for the Vue ecosystem is a new stable (4.0) version of the package vue-router, which came out on December 7th this year. It’s compatible with Vue 3 and works with TypeScript.

**Reply by shaunnbarron** [reply]
> An important piece of information for the Vue ecosystem is a new stable (4.0) version of the package vue-router, which came out on December 7th this year. It’s compatible with Vue 3 and works with TypeScript.

vue-router 4.0 came out in 2020. https://github.com/vuejs/router/releases/tag/v4.0.0

**Comment by nickserv** [comment]
I'm surprised [`URLPattern`](https://wicg.github.io/urlpattern/) hasn't been mentioned in this thread yet. It's still a draft, but it's similar to `path-to-regexp` which React Router already uses. If we can get `URLPattern` working in React Router or Remix it should be easier to have framework agnostic routes.

**Reply by sergiodxa** [reply]
You can already use Remix Router package for a framework agnostic router, RR JS a React wrapper of Remix Router

**Reply by brophdawg11** [reply]
For reference: https://github.com/brophdawg11/remix-routers

**Comment by wobsoriano** [comment]
Fun project https://numix.vercel.app 😆

**Reply by brophdawg11** [reply]
Nice!  I've never gotten deep into the Nuxt module architecture, but I wrote my own Vue+Remix POC some time ago: https://github.com/brophdawg11/vuemix.  Cool to see that you can get the basics working via a module!

**Reply by wobsoriano** [reply]
That looks great! Thinking of copying your PoC and make a module out of it 😁

**Reply by brophdawg11** [reply]
Take that POC with a grain of salt 😉 .  It was super quickly hacked together - and done before I joined the team so I had a super limited understanding of the internals at the time.  It's go no race-condition/cancellation/interruption handling etc.

**Comment by brophdawg11** [comment]
This discussion is being closed as part of the Remix v3 repository cleanup.

Remix v1/v2 was [upstreamed into React Router v7](https://remix.run/blog/wake-up-remix), so if this proposal is still relevant, please reopen or continue the discussion in the [React Router](https://github.com/remix-run/react-router) repository.
